### PR TITLE
refactor(moment): Replace Moment.js with DayJs

### LINKED ISSUE #31178: Remove Moment.js in favor of Day.js
Moment is a legacy library as mentioned in their GitHub repo https://github.com/moment/moment

> Moment.js is a legacy project, now in maintenance mode. In most cases, you should choose a different library.

We have seen issues arising on the [DatePicker](https://github.com/apache/superset/pull/31019) component related to `Moment.js` which led us to the decision to remove `Moment.js` from the repo in favor of `Day.js`. 

`Day.js` offers a much smaller bundle size, it is supported by Ant Design out of the box and it is already installed in Superset.

### PR DESCRIPTION
**Pull request by msyavuz**
### SUMMARY

This pr aims to replace Moment.js with Day.js as the former is a legacy project and is in maintenance mode. There were some design decisions made:

1. Datepicker component in Ant Design v4 accepts Moment object as props. There are couple options to remedy that:
 - Use a temporary plugin to allow Datepicker v4 to accept Day.js.
 - Refactor the Datepicker to allow Day.js.
 - Mark the usages to complete Day.js migration in the Datepicker pr #31019
 
I decided on the last option since other two seemed overkill for a temporary hurdle.

2. Extensions
Day.js object needed to be extended to support our usages. I centralized that in `extendedDayjs` to prevent confusion. Alternative to this is to extend where its used.

### TESTING INSTRUCTIONS

Run the testing suite.

### ADDITIONAL INFORMATION

- [x] Has associated issue: Fixes #31178
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Comment by rusackas** [comment]
It sounds like Ant Design v4 is one hangup here. Pinging @geido since we're in the process of upgrading from v4 to v5, one component at a time. Would v5's Datepicker component mitigate this issue? If so, perhaps we can prioritize that, to get this PR through more easily.

**Comment by msyavuz** [comment]
@rusackas This PR and the Datepicker PR kind of depend on each other since Ant design v4 uses Moment and v5 uses Dayjs. In my opinion finishing the Moment to Dayjs migration on datepicker PR is the most straightforward way.

File: superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberPeriodOverPeriod/transformProps.ts [hunk 1]
```diff
@@ -27,9 +28,11 @@ import {
 } from '@superset-ui/core';
 import { getComparisonFontSize, getHeaderFontSize } from './utils';
 
+dayjs.extend(utc);
+
 export const parseMetricValue = (metricValue: number | string | null) => {
   if (typeof metricValue === 'string') {
-    const dateObject = moment.utc(metricValue, moment.ISO_8601, true);
+    const dateObject = dayjs.utc(metricValue, undefined, true);
```
Comment (geido) [inline superset-frontend/plugins/plugin-chart-echarts/src/BigNumber/BigNumberPeriodOverPeriod/transformProps.ts:35]: Just so I can understand the scope of the change here, what is `undefined` doing here?

File: superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx [hunk 2]
```diff
@@ -46,20 +41,20 @@ const openSelectMenu = () => {
...
 
-test('use the timezone from `moment` if no timezone provided', async () => {
+test('use the timezone from `dayjs` if no timezone provided', async () => {
   const TimezoneSelector = await loadComponent('2022-01-01');
   const onTimezoneChange = jest.fn();
   render(<TimezoneSelector onTimezoneChange={onTimezoneChange} />);
-  expect(screen.getByLabelText('Loading')).toBeVisible();
-  await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
+  // expect(screen.getByLabelText('Loading')).toBeVisible();
+  // await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
```
Comment (geido) [inline superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx:55]: Do we need to remove these? If so, what's causing this to be inapplicable?

File: superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx [hunk 3]
```diff
@@ -71,9 +66,9 @@ test('update to closest deduped timezone when timezone is provided', async () =>
       timezone="America/Los_Angeles"
     />,
   );
-  await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
+  // await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
```
Comment (geido) [inline superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx:69]: Same as my other comment

File: superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx [hunk 4]
```diff
@@ -113,17 +106,17 @@ test('can select a timezone values and returns canonical timezone name', async (
       timezone="Africa/Abidjan"
     />,
   );
-  await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
+  // await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
```
Comment (geido) [inline superset-frontend/src/components/TimezoneSelector/TimezoneSelector.test.tsx:109]: One more

File: superset-frontend/src/components/TimezoneSelector/index.tsx [hunk 5]
```diff
@@ -82,43 +70,40 @@ export default function TimezoneSelector({
...
             ? offsetsToName[offsets]?.[1]
             : offsetsToName[offsets]?.[0]) || name
         );
       };
 
-      const ALL_ZONES = momentLib.tz
-        .countries()
-        .map(country => momentLib.tz.zonesForCountry(country, true))
-        .flat();
+      // TODO: remove this ts-ignore when typescript is upgraded to 5.1
```
Comment (geido) [inline superset-frontend/src/components/TimezoneSelector/index.tsx:79]: What's the TLDR?

File: superset-frontend/src/features/alerts/AlertReportModal.test.tsx [hunk 6]
```diff
@@ -526,7 +523,9 @@ test('renders default Schedule fields', async () => {
     useRedux: true,
   });
   userEvent.click(screen.getByTestId('schedule-panel'));
-  await waitForElementToBeRemoved(() => screen.queryByLabelText('Loading'));
+  if (screen.queryByLabelText('Loading')) {
```
Comment (geido) [inline superset-frontend/src/features/alerts/AlertReportModal.test.tsx:526]: I think conditionals in tests might not be the preferred approach. What's the scenario?

File: superset-frontend/src/pages/AnnotationList/index.tsx [hunk 7]
```diff
@@ -170,7 +170,7 @@ function AnnotationList({
           row: {
             original: { start_dttm: startDttm },
           },
-        }: any) => moment(new Date(startDttm)).format('ll'),
+        }: any) => dayjs(new Date(startDttm)).format('ll'),
```
Comment (geido) [inline superset-frontend/src/pages/AnnotationList/index.tsx:173]: A good time to remove this `any` for a better type?

File: superset-frontend/src/pages/AnnotationList/index.tsx [hunk 8]
```diff
@@ -179,7 +179,7 @@ function AnnotationList({
           row: {
             original: { end_dttm: endDttm },
           },
-        }: any) => moment(new Date(endDttm)).format('ll'),
+        }: any) => dayjs(new Date(endDttm)).format('ll'),
```
Comment (geido) [inline superset-frontend/src/pages/AnnotationList/index.tsx:182]: Same comment about removing `any`

File: superset-frontend/src/pages/ExecutionLogList/index.tsx [hunk 9]
```diff
@@ -119,7 +119,7 @@ function ExecutionLog({
             original: { scheduled_dttm: scheduledDttm },
           },
         }: any) =>
```
Comment (geido) [inline superset-frontend/src/pages/ExecutionLogList/index.tsx:121]: Can we now use a better type than `any`?

File: superset-frontend/src/pages/ExecutionLogList/index.tsx [hunk 10]
```diff
@@ -128,7 +128,7 @@ function ExecutionLog({
           row: {
             original: { start_dttm: startDttm },
           },
-        }: any) => moment(new Date(startDttm)).format('YYYY-MM-DD hh:mm:ss a'),
+        }: any) => dayjs(new Date(startDttm)).format('YYYY-MM-DD hh:mm:ss a'),
```
Comment (geido) [inline superset-frontend/src/pages/ExecutionLogList/index.tsx:131]: One more about `any`

**Review by geido — COMMENTED** [review]
Great job! A few nits

File: superset-frontend/src/features/alerts/AlertReportModal.test.tsx (hunk 6 above)
Comment (msyavuz) [reply superset-frontend/src/features/alerts/AlertReportModal.test.tsx:526]: Loading no longer exists so removed this as well

File: superset-frontend/src/pages/ExecutionLogList/index.tsx (hunk 9 above)
Comment (msyavuz) [reply superset-frontend/src/pages/ExecutionLogList/index.tsx:122]: For this one i think keeping any is ok still

**Review by geido — APPROVED** [review]

**Comment by tahvane1** [comment]
Why timezones are removed? (dedup...) This broke timezones in alerts & reports...

**Comment by msyavuz** [comment]
@tahvane1 Hey, can you explain what exactly is broken in alerts and reports?

**Pull request merged**